static/download-service.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2023, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import $ from 'jquery';
+
+import {FiledataPair} from '../types/compilation/compilation.interfaces';
+
+export type SourceAndFiles = {
+    files: FiledataPair[];
+    source: string;
+};
+
+export type FileToDownload = {
+    url: string;
+    filename: string;
+    contents?: string;
+};
+
+export class IncludeDownloads {
+    private toDownload: FileToDownload[] = [];
+    private downloadPromises: Promise<FileToDownload>[] = [];
+
+    private async startDownload(download) {
+        this.downloadPromises.push(
+            new Promise(resolve => {
+                $.get(download.url, data => {
```

Comment:
what about errors? what happens then? where does the exception end up?

Reply:
Good question, no idea, will need to test

Reply:
I _wish_ we could: https://caniuse.com/fetch paints a not-so-nice picture :(

Reply:
Disregard this, it actually looks great

Reply:
We can already use fetch since we have a polyfill for it :) https://github.com/compiler-explorer/compiler-explorer/commit/85a287c011c7e9b02f8d55ede3f6ee1d40c503df

https://github.com/compiler-explorer/compiler-explorer/blob/main/static/main.ts#L28

Reply:
replaced $.get with fetch

Reply:
also added error handling, in case of a failed download, it will still include the url or something from the error, and then you'll see a compilation error hinting at what happened